### PR TITLE
Reorganize parsers

### LINKED ISSUE #21: Handle webdriver timeouts
Hello,
I am trying to use protractor-flake to rerun some failed scenarios that are failing sometimes because my selenium grid gives a timeout when forwards some tests to the nodes.
When this occurs, I get this log:

`03:43:15.738 [chrome #01-34] WebDriverError: Error forwarding the new session Error forwarding the request Read timed out
03:43:15.738 [chrome #01-34]     at WebDriverError (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/error.js:26:26)
03:43:15.738 [chrome #01-34]     at Object.checkLegacyResponse (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/error.js:580:13)
03:43:15.738 [chrome #01-34]     at /var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/webdriver.js:64:13
03:43:15.738 [chrome #01-34]     at Promise.invokeCallback_ (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/promise.js:1329:14)
03:43:15.738 [chrome #01-34]     at TaskQueue.execute_ (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/promise.js:2790:14)
03:43:15.738 [chrome #01-34]     at TaskQueue.executeNext_ (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/promise.js:2773:21)
03:43:15.738 [chrome #01-34]     at /var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/promise.js:2652:27
03:43:15.738 [chrome #01-34]     at /var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/promise.js:639:7
03:43:15.738 [chrome #01-34]     at process._tickCallback (node.js:405:9)
03:43:15.738 [chrome #01-34] From: Task: WebDriver.createSession()
03:43:15.738 [chrome #01-34]     at acquireSession (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/webdriver.js:62:22)
03:43:15.739 [chrome #01-34]     at Function.createSession (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/lib/webdriver.js:295:12)
03:43:15.739 [chrome #01-34]     at Builder.build (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/selenium-webdriver/builder.js:458:24)
03:43:15.739 [chrome #01-34]     at [object Object].DriverProvider.getNewDriver (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/protractor/built/driverProviders/driverProvider.js:42:27)
03:43:15.739 [chrome #01-34]     at [object Object].Runner.createBrowser (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/protractor/built/runner.js:203:37)
03:43:15.739 [chrome #01-34]     at /var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/protractor/built/runner.js:293:21
03:43:15.739 [chrome #01-34]     at _fulfilled (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/q/q.js:834:54)
03:43:15.739 [chrome #01-34]     at self.promiseDispatch.done (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/q/q.js:863:30)
03:43:15.739 [chrome #01-34]     at Promise.promise.promiseDispatch (/var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/q/q.js:796:13)
03:43:15.739 [chrome #01-34]     at /var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/node_modules/q/q.js:556:49`

So on the final log, I have checked that these type of tests (failed with exit code 1) are not rerunned.

`03:19:06.063 [launcher] chrome #01-6 failed with exit code: 1
03:19:06.063 [launcher] chrome #01-12 passed
03:19:06.064 [launcher] chrome #01-13 passed
03:19:06.064 [launcher] chrome #01-14 passed
03:19:06.064 [launcher] chrome #01-11 failed with exit code: 1
03:19:06.064 [launcher] chrome #01-15 passed
03:19:06.064 [launcher] chrome #01-16 passed
03:19:06.064 [launcher] chrome #01-17 passed
03:19:06.064 [launcher] chrome #01-18 passed
03:19:06.064 [launcher] chrome #01-19 passed
03:19:06.064 [launcher] chrome #01-20 passed
03:19:06.064 [launcher] chrome #01-21 passed
03:19:06.064 [launcher] chrome #01-23 passed
03:19:06.064 [launcher] chrome #01-22 failed with exit code: 1
03:19:06.064 [launcher] chrome #01-25 passed
03:19:06.064 [launcher] chrome #01-24 failed with exit code: 1
03:19:06.064 [launcher] chrome #01-26 passed
03:19:06.064 [launcher] chrome #01-27 passed
03:19:06.064 [launcher] chrome #01-28 passed
03:19:06.064 [launcher] chrome #01-29 passed
03:19:06.065 [launcher] chrome #01-30 passed
03:19:06.065 [launcher] chrome #01-31 passed
03:19:06.065 [launcher] chrome #01-33 passed
03:19:06.065 [launcher] chrome #01-34 failed 1 test(s)
03:19:06.065 [launcher] chrome #01-35 failed 1 test(s)
03:19:06.065 [launcher] chrome #01-36 passed
03:19:06.065 [launcher] chrome #01-32 failed with exit code: 1
03:19:06.065 [launcher] overall: 2 failed spec(s) and 5 process(es) failed to complete
03:19:06.065 [launcher] Process exited with error code 100
03:19:06.072 Re-running tests: test attempt 2
03:19:06.072 Re-running the following test files:
03:19:06.072 /var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/mock-tests/tests/account/desktop/my-tracks-1-spec.js
03:19:06.072 /var/lib/go-agent-4/pipelines/FrontEnd-TVG4-Dev/workspace/tvg-tvg4-angular/tests/web-tests/mock-tests/tests/account/desktop/my-tracks-2-spec.js`

The only ones that are re-runned are these -> (03:19:06.065 [launcher] chrome #01-34 failed 1 test(s)).

Please edit this comment to see the breaklines.

### PR DESCRIPTION
**Pull request by NickTomlin**
This PR (unfortunately) does a lot: 
1. Introduce the ability to parse multi test output for failures even if the test name is not included in the parsed output (should fix #21). This feature is thanks to the work of @FinKingma
2. Split parsers into separate modules (instead of a giant `if/else if/else`
3. Reorganize tests and make them "nicer" (by shortening urls and cleaning up file structure)

I'm going to let this sit out for a little while before merging; this should go out as a breaking release.

@ayush @FinKingma would appreciate some 👀  on this. Thanks!

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,30 +1,30 @@
...
 
-  function handleTestEnd (status, output) {
+  function handleTestEnd (status, output = '') {
     if (status === 0) {
       callback(status)
     } else {
       if (++testAttempt <= parsedOptions.maxAttempts) {
-        let failedSpecs = failedSpecParser(output)
+        let parser = getParser(parsedOptions.parser, output)
+        log('info', `Using ${parser.name} to parse output`)
```
Comment (FinKingma) [inline src/index.js]: you're missing a \n here.
Should be something like: 
`log('info', 'Using ${parser.name} to parse output\n')`

File: src/parsers/index.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+function getParser (name, output = '') {
+  let parser
+  if (name) {
+    parser = all.find((parser) => name === parser.name)
+    if (!parser) { throw new Error(`Invalid Parser Specified: ${name}`) }
+  } else {
+    parser = all.find((p) => {
+      return p.test(output)
+    })
```
Comment (FinKingma) [inline src/parsers/index.js]: Got some doubts about this last part (trying to discover the correct parser when none was specified). It is very unclear to users how this is done (magic). I know that protractor-flake did this before, but now that we have another parser (total 3) I propose a cleaner solution: we select standardParser by default and let the user decide which parser to use. This will make the break in the change a worse (for cucumber users), but I do think this is the cleanest option.

If you still want to keep this automatic detection in protractor-flake, I propose having the test function of multitestParser return false by default. This might even turn this change in non-breaking, because the multitestParser will never be used by default.

So in short I see 2 options:
1) make the break worse, but reduce the magic done by selecting standardParser by default.
2) fix the break by never selecting multitestParser by default (have it's test function return false by default).

Comment (NickTomlin) [inline src/parsers/index.js]: Great points! I think i'm in agreement on this; it's trying too hard to make things easy. Most people who are going to be using this will know what kind of output they will be using and it's not too heavy to set this up once.

I think 1. is the best and implemented the following way: 
1. Remove all test detection
2. Default `parser` option to `standard`
3. Allow users to select a parser with the `parser` option
4. Document the following in the README

⭐ bonus⭐  
Allow users to specify their own parser module which protractor-flake will require and use?

File: src/index.js (hunk 1 above)
Comment (NickTomlin) [inline src/index.js]: Ah yep! I wonder if it would be best to just add a check to the logger that sees if there's a `\n` at the end or not, that way it would work seamlessly with stdout from protractor.

Comment (FinKingma) [inline src/index.js]: That would be great. Perhaps the check won't even be necessary, you can add a \n at the end of each logline.

File: src/parsers/index.js (hunk 2 above)
Comment (FinKingma) [inline src/parsers/index.js]: You mean creating a new header for the parsers?
I was thinking more along the lines of adding the parser option in the usage header:

Hmm that kinda sounds like users have to write their own parser module XD

How about

⭐Parsers⭐
The following parsers can be used to discover and rerun failed tests.
- StandardParser: will parse stack trace error messages and rerun the testfiles that resulted in these error.
- CucumberParser: will parse cucumber test output and rerun the failed tests.
- multiTestParser: will parse multi-capabilities or sharded test file output and rerun the failed tests.

And then also adding the parser option to the 'usage' header.

**Comment by NickTomlin** [comment]
@FinKingma I'm going to move this discussion out of comments for ease of use. 

I've updated to accept a `parser` option and default to the `standard` parser. (I think we were both on the same page feature wise).

**Comment by NickTomlin** [comment]
Closed and released as `2.0.1` due to a semantic-release flub. This was originally released as `1.1.0` but that version has been unpublished because it contained breaking changes.
